File: mobile/android/app/src/main/java/com/xfshopee/JavaScriptBridge.kt
package com.xfshopee

import android.annotation.SuppressLint
import android.content.Context
import android.webkit.JavascriptInterface
import android.webkit.WebView
import org.json.JSONObject
import java.util.*

/**
 * JavaScript桥接接口
 * 
 * 提供原生功能给WebView中的JavaScript调用
 * 双向通信：JavaScript调用原生方法，原生发送事件到JavaScript
 */
@SuppressLint("JavascriptInterface")
class JavaScriptBridge(
    private val context: Context,
    private val webViewManager: WebViewManager
) {
    
    companion object {
        // 消息ID生成器
        private var messageIdCounter = 0
            get() {
                field++
                return field
            }
        
        // 回调映射：callbackId -> 回调函数
        private val callbacks: MutableMap<String, (String) -> Unit> = mutableMapOf()
    }
    
    // 事件监听器：eventName -> 监听器列表
    private val eventListeners: MutableMap<String, MutableList<(JSONObject) -> Unit>> = mutableMapOf()
    
    // 存储管理器
    private val storageManager: StorageManager by lazy {
        StorageManager(context)
    }
    
    // 扫码管理器
    private val scannerManager: ScannerManager by lazy {
        ScannerManager(context)
    }
    
    // Cookie管理器
    private val cookieManager: CookieManager by lazy {
        com.xfshopee.CookieManager(context)
    }
    
    /**
     * JavaScript调用原生方法的主入口
     */
    @JavascriptInterface
    fun callNative(messageJson: String): String {
        return try {
            val message = JSONObject(messageJson)
            val messageId = message.optString("messageId", "")
            val action = message.optString("action", "")
            val params = message.optJSONObject("params") ?: JSONObject()
            val callbackId = message.optString("callbackId", "")
            
            // 处理请求
            val result = handleAction(action, params)
            
            // 构建响应
            val response = JSONObject().apply {
                put("messageId", messageId)
                put("success", true)
                put("data", result)
                put("error", JSONObject.NULL)
                put("timestamp", System.currentTimeMillis())
            }
            
            response.toString()
        } catch (e: Exception) {
            // 错误响应
            val errorResponse = JSONObject().apply {
                put("messageId", "error_${System.currentTimeMillis()}")
                put("success", false)
                put("data", JSONObject.NULL)
                put("error", JSONObject().apply {
                    put("code", "BRIDGE_ERROR")
                    put("message", e.message ?: "Unknown error")
                    put("details", JSONObject())
                })
                put("timestamp", System.currentTimeMillis())
            }
            
            errorResponse.toString()
        }
    }
    
    /**
     * 处理具体动作
     */
    private fun handleAction(action: String, params: JSONObject): JSONObject {
        return when (action) {
            // 存储相关
            "storage.set" -> handleStorageSet(params)
            "storage.get" -> handleStorageGet(params)
            "storage.remove" -> handleStorageRemove(params)
            "storage.clear" -> handleStorageClear(params)
            
            // 扫码相关
            "scanner.startScan" -> handleScannerStart(params)
            "scanner.scanFromImage" -> handleScanFromImage(params)
            
            // Cookie相关
            "cookie.get" -> handleCookieGet(params)
            "cookie.set" -> handleCookieSet(params)
            "cookie.clear" -> handleCookieClear(params)
            
            // 设备信息
            "device.getInfo" -> handleDeviceInfo(params)
            "device.getNetworkStatus" -> handleNetworkStatus(params)
            
            // 应用功能
            "app.getConfig" -> handleAppConfig(params)
            "app.openModule" -> handleOpenModule(params)
            "app.closeModule" -> handleCloseModule(params)
            
            // 文件操作
            "file.read" -> handleFileRead(params)
            "file.write" -> handleFileWrite(params)
            "file.delete" -> handleFileDelete(params)
            
            // 其他
            else -> JSONObject().apply {
                put("error", "ACTION_NOT_SUPPORTED")
                put("action", action)
            }
        }
    }
    
    /**
     * 存储：设置值
     */
    private fun handleStorageSet(params: JSONObject): JSONObject {
        val key = params.optString("key", "")
        val value = params.optString("value", "")
        
        if (key.isNotEmpty()) {
            storageManager.setString(key, value)
        }
        
        return JSONObject().apply {
            put("success", true)
        }
    }
    
    /**
     * 存储：获取值
     */
    private fun handleStorageGet(params: JSONObject): JSONObject {
        val key = params.optString("key", "")
        val defaultValue = params.optString("defaultValue", "")
        
        val value = if (key.isNotEmpty()) {
            storageManager.getString(key, defaultValue)
        } else {
            defaultValue
        }
        
        return JSONObject().apply {
            put("value", value)
            put("success", true)
        }
    }
    
    /**
     * 存储：移除值
     */
    private fun handleStorageRemove(params: JSONObject): JSONObject {
        val key = params.optString("key", "")
        
        if (key.isNotEmpty()) {
            storageManager.remove(key)
        }
        
        return JSONObject().apply {
            put("success", true)
        }
    }
    
    /**
     * 存储：清空
     */
    private fun handleStorageClear(params: JSONObject): JSONObject {
        storageManager.clear()
        return JSONObject().apply {
            put("success", true)
        }
    }
    
    /**
     * 扫码：开始扫描
     */
    private fun handleScannerStart(params: JSONObject): JSONObject {
        val format = params.optString("format", "QR_CODE")
        
        return try {
            val result = scannerManager.scan(format)
            JSONObject().apply {
                put("success", true)
                put("data", result.data)
                put("format", result.format)
                put("timestamp", result.timestamp)
            }
        } catch (e: Exception) {
            JSONObject().apply {
                put("success", false)
                put("error", e.message ?: "Scan failed")
            }
        }
    }
    
    /**
     * 扫码：从图片扫描
     */
    private fun handleScanFromImage(params: JSONObject): JSONObject {
        val imageData = params.optString("imageData", "")
        
        if (imageData.isEmpty()) {
            return JSONObject().apply {
                put("success", false)
                put("error", "No image data provided")
            }
        }
        
        return try {
            val result = scannerManager.scanFromImage(imageData)
            JSONObject().apply {
                put("success", true)
                put("data", result.data)
                put("format", result.format)
                put("timestamp", result.timestamp)
            }
        } catch (e: Exception) {
            JSONObject().apply {
                put("success", false)
                put("error", e.message ?: "Image scan failed")
            }
        }
    }
    
    /**
     * Cookie：获取
     */
    private fun handleCookieGet(params: JSONObject): JSONObject {
        val domain = params.optString("domain", "a.b.c")
        val cookies = cookieManager.getCookies(domain)
        
        return JSONObject().apply {
            put("success", true)
            put("cookies", cookies.joinToString("; "))
        }
    }
    
    /**
     * Cookie：设置
     */
    private fun handleCookieSet(params: JSONObject): JSONObject {
        val cookieString = params.optString("cookie", "")
        val domain = params.optString("domain", "a.b.c")
        
        if (cookieString.isNotEmpty()) {
            cookieManager.setCookie(domain, cookieString)
        }
        
        return JSONObject().apply {
            put("success", true)
        }
    }
    
    /**
     * Cookie：清空
     */
    private fun handleCookieClear(params: JSONObject): JSONObject {
        val domain = params.optString("domain", "")
        
        if (domain.isNotEmpty()) {
            cookieManager.clearCookies(domain)
        } else {
            cookieManager.clearAllCookies()
        }
        
        return JSONObject().apply {
            put("success", true)
        }
    }
    
    /**
     * 设备：获取信息
     */
    private fun handleDeviceInfo(params: JSONObject): JSONObject {
        return JSONObject().apply {
            put("platform", "Android")
            put("osVersion", android.os.Build.VERSION.RELEASE)
            put("deviceModel", android.os.Build.MODEL)
            put("appVersion", "1.0.0")
            put("screenWidth", context.resources.displayMetrics.widthPixels)
            put("screenHeight", context.resources.displayMetrics.heightPixels)
            put("success", true)
        }
    }
    
    /**
     * 设备：获取网络状态
     */
    private fun handleNetworkStatus(params: JSONObject): JSONObject {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as android.net.ConnectivityManager
        val networkInfo = connectivityManager.activeNetworkInfo
        
        return JSONObject().apply {
            put("connected", networkInfo?.isConnected == true)
            put("type", networkInfo?.typeName ?: "UNKNOWN")
            put("success", true)
        }
    }
    
    /**
     * 应用：获取配置
     */
    private fun handleAppConfig(params: JSONObject): JSONObject {
        return JSONObject().apply {
            put("baseUrl", "http://a.b.c")
            put("apiPrefix", "/xfbh/api/v1")
            put("mobilePrefix", "/xfbh/mobile")
            put("environment", "production")
            put("success", true)
        }
    }
    
    /**
     * 应用：打开模块
     */
    private fun handleOpenModule(params: JSONObject): JSONObject {
        val moduleId = params.optString("moduleId", "")
        // 实际打开模块的逻辑在MainActivity中处理
        // 这里只是返回成功
        return JSONObject().apply {
            put("success", true)
            put("moduleId", moduleId)
        }
    }
    
    /**
     * 应用：关闭模块
     */
    private fun handleCloseModule(params: JSONObject): JSONObject {
        val moduleId = params.optString("moduleId", "")
        // 实际关闭模块的逻辑在WebViewManager中处理
        return JSONObject().apply {
            put("success", true)
            put("moduleId", moduleId)
        }
    }
    
    /**
     * 文件：读取
     */
    private fun handleFileRead(params: JSONObject): JSONObject {
        // 简化实现
        return JSONObject().apply {
            put("success", false)
            put("error", "Not implemented")
        }
    }
    
    /**
     * 文件：写入
     */
    private fun handleFileWrite(params: JSONObject): JSONObject {
        // 简化实现
        return JSONObject().apply {
            put("success", false)
            put("error", "Not implemented")
        }
    }
    
    /**
     * 文件：删除
     */
    private fun handleFileDelete(params: JSONObject): JSONObject {
        // 简化实现
        return JSONObject().apply {
            put("success", false)
            put("error", "Not implemented")
        }
    }
    
    /**
     * 发送事件到JavaScript
     */
    fun sendEventToJavaScript(webView: WebView, eventName: String, data: JSONObject) {
        val eventMessage = JSONObject().apply {
            put("type", "event")
            put("event", eventName)
            put("data", data)
            put("timestamp", System.currentTimeMillis())
        }
        
        val javascript = """
            (function() {
                if (window.nativeBridge && window.nativeBridge.onEvent) {
                    window.nativeBridge.onEvent(${eventMessage.toString()});
                }
                window.dispatchEvent(new CustomEvent('nativeEvent', {
                    detail: ${eventMessage.toString()}
                }));
            })();
        """.trimIndent()
        
        // 在主线程执行
        webView.post {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
                webView.evaluateJavascript(javascript, null)
            } else {
                webView.loadUrl("javascript:$javascript")
            }
        }
    }
    
    /**
     * 注册事件监听器
     */
    fun addEventListener(eventName: String, listener: (JSONObject) -> Unit) {
        val listeners = eventListeners.getOrPut(eventName) { mutableListOf() }
        listeners.add(listener)
    }
    
    /**
     * 移除事件监听器
     */
    fun removeEventListener(eventName: String, listener: (JSONObject) -> Unit) {
        eventListeners[eventName]?.remove(listener)
    }
    
    /**
     * 触发事件
     */
    fun triggerEvent(eventName: String, data: JSONObject) {
        eventListeners[eventName]?.forEach { listener ->
            listener(data)
        }
    }
}

/**
 * 扫码结果数据类
 */
data class ScanResult(
    val success: Boolean,
    val data: String,
    val format: String,
    val timestamp: Long = System.currentTimeMillis()
)

File: mobile/ios/Native/JavaScriptBridge.swift
import UIKit
import WebKit

/**
 * JavaScript桥接
 * 
 * 提供原生功能给WebView中的JavaScript调用
 */
class JavaScriptBridge: NSObject {
    
    // 单例实例
    static let shared = JavaScriptBridge()
    
    // 事件监听器：eventName -> 监听器列表
    private var eventListeners: [String: [(Any) -> Void]] = [:]
    
    // 存储管理器
    private let storageManager = StorageManager.shared
    
    // 扫码管理器
    private let scannerManager = ScannerManager.shared
    
    // Cookie管理器
    private let cookieManager = CookieManager.shared
    
    private override init() {
        super.init()
    }
    
    /**
     * 注入JavaScript桥接到WebView
     */
    func inject(to webView: WKWebView) {
        // 创建用户脚本
        let bridgeScript = WKUserScript(
            source: injectionScript,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: true
        )
        
        webView.configuration.userContentController.addUserScript(bridgeScript)
        
        // 添加消息处理器
        webView.configuration.userContentController.add(self, name: "nativeBridge")
    }
    
    /**
     * JavaScript注入脚本
     */
    private var injectionScript: String {
        return """
        (function() {
            // 创建NativeBridge对象
            window.NativeBridge = {
                // 调用原生方法
                callNative: function(message) {
                    return new Promise((resolve, reject) => {
                        const messageId = 'msg_' + Date.now() + '_' + Math.random().toString(36).substr(2, 9);
                        
                        // 发送消息到原生
                        window.webkit.messageHandlers.nativeBridge.postMessage({
                            ...message,
                            messageId: messageId
                        });
                        
                        // 设置回调
                        window['__native_callback_' + messageId] = { resolve, reject };
                        
                        // 超时处理
                        setTimeout(() => {
                            if (window['__native_callback_' + messageId]) {
                                delete window['__native_callback_' + messageId];
                                reject(new Error('Timeout'));
                            }
                        }, 30000);
                    });
                },
                
                // 添加事件监听器
                addEventListener: function(eventName, callback) {
                    if (!window.__native_event_listeners) {
                        window.__native_event_listeners = {};
                    }
                    if (!window.__native_event_listeners[eventName]) {
                        window.__native_event_listeners[eventName] = [];
                    }
                    window.__native_event_listeners[eventName].push(callback);
                },
                
                // 移除事件监听器
                removeEventListener: function(eventName, callback) {
                    if (!window.__native_event_listeners || !window.__native_event_listeners[eventName]) {
                        return;
                    }
                    const index = window.__native_event_listeners[eventName].indexOf(callback);
                    if (index !== -1) {
                        window.__native_event_listeners[eventName].splice(index, 1);
                    }
                }
            };
            
            // 触发事件给JavaScript
            window.dispatchNativeEvent = function(eventName, data) {
                if (!window.__native_event_listeners || !window.__native_event_listeners[eventName]) {
                    return;
                }
                window.__native_event_listeners[eventName].forEach(callback => {
                    try {
                        callback(data);
                    } catch (error) {
                        console.error('Error in event listener:', error);
                    }
                });
            };
            
            console.log('Native Bridge injected');
        })();
        """
    }
    
    /**
     * 发送事件到JavaScript
     */
    func sendEvent(to webView: WKWebView, eventName: String, data: Any) {
        let script = """
        if (window.dispatchNativeEvent) {
            window.dispatchNativeEvent('\(eventName)', \(jsonString(from: data)));
        }
        """
        
        webView.evaluateJavaScript(script) { result, error in
            if let error = error {
                print("Failed to send event to JavaScript: \(error)")
            }
        }
    }
    
    /**
     * 处理来自JavaScript的消息
     */
    private func handleMessage(_ message: [String: Any], completion: @escaping (Any) -> Void) {
        guard let action = message["action"] as? String else {
            completion(createErrorResponse(messageId: message["messageId"] as? String, error: "Missing action"))
            return
        }
        
        let params = message["params"] as? [String: Any] ?? [:]
        let messageId = message["messageId"] as? String ?? ""
        
        // 处理动作
        let result: Any
        switch action {
        case "storage.set":
            result = handleStorageSet(params)
        case "storage.get":
            result = handleStorageGet(params)
        case "storage.remove":
            result = handleStorageRemove(params)
        case "storage.clear":
            result = handleStorageClear(params)
        case "scanner.startScan":
            result = handleScannerStart(params)
        case "scanner.scanFromImage":
            result = handleScanFromImage(params)
        case "cookie.get":
            result = handleCookieGet(params)
        case "cookie.set":
            result = handleCookieSet(params)
        case "cookie.clear":
            result = handleCookieClear(params)
        case "device.getInfo":
            result = handleDeviceInfo(params)
        case "device.getNetworkStatus":
            result = handleNetworkStatus(params)
        case "app.getConfig":
            result = handleAppConfig(params)
        default:
            result = createErrorResponse(messageId: messageId, error: "Unsupported action: \(action)")
        }
        
        completion(result)
    }
    
    /**
     * 存储：设置值
     */
    private func handleStorageSet(_ params: [String: Any]) -> [String: Any] {
        guard let key = params["key"] as? String,
              let value = params["value"] as? String else {
            return ["success": false, "error": "Missing key or value"]
        }
        
        storageManager.setString(key: key, value: value)
        return ["success": true]
    }
    
    /**
     * 存储：获取值
     */
    private func handleStorageGet(_ params: [String: Any]) -> [String: Any] {
        guard let key = params["key"] as? String else {
            return ["success": false, "error": "Missing key"]
        }
        
        let defaultValue = params["defaultValue"] as? String ?? ""
        let value = storageManager.getString(key: key, defaultValue: defaultValue)
        
        return ["success": true, "value": value]
    }
    
    /**
     * 存储：移除值
     */
    private func handleStorageRemove(_ params: [String: Any]) -> [String: Any] {
        guard let key = params["key"] as? String else {
            return ["success": false, "error": "Missing key"]
        }
        
        storageManager.remove(key: key)
        return ["success": true]
    }
    
    /**
     * 存储：清空
     */
    private func handleStorageClear(_ params: [String: Any]) -> [String: Any] {
        storageManager.clear()
        return ["success": true]
    }
    
    /**
     * 扫码：开始扫描
     */
    private func handleScannerStart(_ params: [String: Any]) -> [String: Any] {
        let format = params["format"] as? String ?? "QR_CODE"
        
        // 注意：这里需要在实际应用中实现扫码逻辑
        // 暂时返回模拟数据
        return [
            "success": true,
            "data": format == "QR_CODE" ? "https://xfshopee.com/product/123" : "123456789012",
            "format": format,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]
    }
    
    /**
     * 扫码：从图片扫描
     */
    private func handleScanFromImage(_ params: [String: Any]) -> [String: Any] {
        guard let imageData = params["imageData"] as? String else {
            return ["success": false, "error": "Missing image data"]
        }
        
        // 注意：这里需要在实际应用中实现图片扫码逻辑
        return [
            "success": false,
            "error": "Not implemented"
        ]
    }
    
    /**
     * Cookie：获取
     */
    private func handleCookieGet(_ params: [String: Any]) -> [String: Any] {
        let domain = params["domain"] as? String ?? "a.b.c"
        let cookies = cookieManager.getCookies(for: domain)
        
        return [
            "success": true,
            "cookies": cookies.joined(separator: "; ")
        ]
    }
    
    /**
     * Cookie：设置
     */
    private func handleCookieSet(_ params: [String: Any]) -> [String: Any] {
        guard let cookie = params["cookie"] as? String else {
            return ["success": false, "error": "Missing cookie"]
        }
        
        let domain = params["domain"] as? String ?? "a.b.c"
        cookieManager.setCookie(cookie, for: domain)
        
        return ["success": true]
    }
    
    /**
     * Cookie：清空
     */
    private func handleCookieClear(_ params: [String: Any]) -> [String: Any] {
        let domain = params["domain"] as? String
        
        if let domain = domain {
            cookieManager.clearCookies(for: domain)
        } else {
            cookieManager.clearAllCookies()
        }
        
        return ["success": true]
    }
    
    /**
     * 设备：获取信息
     */
    private func handleDeviceInfo(_ params: [String: Any]) -> [String: Any] {
        let device = UIDevice.current
        let screen = UIScreen.main
        
        return [
            "success": true,
            "platform": "iOS",
            "osVersion": device.systemVersion,
            "deviceModel": device.model,
            "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0",
            "screenWidth": Int(screen.bounds.width * screen.scale),
            "screenHeight": Int(screen.bounds.height * screen.scale)
        ]
    }
    
    /**
     * 设备：获取网络状态
     */
    private func handleNetworkStatus(_ params: [String: Any]) -> [String: Any] {
        // 注意：需要导入SystemConfiguration框架
        // 暂时返回模拟数据
        return [
            "success": true,
            "connected": true,
            "type": "wifi"
        ]
    }
    
    /**
     * 应用：获取配置
     */
    private func handleAppConfig(_ params: [String: Any]) -> [String: Any] {
        return [
            "success": true,
            "baseUrl": "http://a.b.c",
            "apiPrefix": "/xfbh/api/v1",
            "mobilePrefix": "/xfbh/mobile",
            "environment": "production"
        ]
    }
    
    /**
     * 创建错误响应
     */
    private func createErrorResponse(messageId: String?, error: String) -> [String: Any] {
        return [
            "messageId": messageId ?? "",
            "success": false,
            "data": NSNull(),
            "error": [
                "code": "BRIDGE_ERROR",
                "message": error,
                "details": [:]
            ],
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]
    }
    
    /**
     * 转换为JSON字符串
     */
    private func jsonString(from object: Any) -> String {
        do {
            let data = try JSONSerialization.data(withJSONObject: object, options: [])
            return String(data: data, encoding: .utf8) ?? "null"
        } catch {
            return "null"
        }
    }
}

// MARK: - WKScriptMessageHandler
extension JavaScriptBridge: WKScriptMessageHandler {
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == "nativeBridge",
              let messageBody = message.body as? [String: Any] else {
            return
        }
        
        // 处理消息
        handleMessage(messageBody) { result in
            // 发送响应回JavaScript
            if let webView = message.webView {
                self.sendResponse(to: webView, messageId: messageBody["messageId"] as? String, result: result)
            }
        }
    }
    
    /**
     * 发送响应回JavaScript
     */
    private func sendResponse(to webView: WKWebView, messageId: String?, result: Any) {
        var response: [String: Any] = [
            "messageId": messageId ?? "",
            "success": true,
            "data": result,
            "error": NSNull(),
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]
        
        // 如果result本身已经包含错误信息，调整响应
        if let resultDict = result as? [String: Any],
           resultDict["success"] as? Bool == false {
            response["success"] = false
            response["data"] = NSNull()
            response["error"] = resultDict
        }
        
        let script = """
        if (window['__native_callback_\(messageId ?? "")']) {
            const callback = window['__native_callback_\(messageId ?? "")'];
            delete window['__native_callback_\(messageId ?? "")'];
            
            callback.resolve(\(jsonString(from: response)));
        }
        """
        
        webView.evaluateJavaScript(script) { _, error in
            if let error = error {
                print("Failed to send response to JavaScript: \(error)")
            }
        }
    }
}